File: src/python/claude_mem/storage/tiered/query_router.py
"""
Query Router

Routes search queries to appropriate storage tier based on query characteristics.
Implements hot-first search with cold tier fallback.
"""

import logging
from typing import Optional

from claude_recall.models import (
    SearchResult,
    RAGQueryResult,
    HybridSearchOptions,
)
from claude_recall.storage.cold.hybrid_search import (
    PostgresObservationSearch,
    PostgresSummarySearch,
)
from claude_recall.storage.hot.vector_search import (
    RedisObservationSearch,
    RedisSummarySearch,
)
from claude_recall.storage.cold.postgres_client import get_postgres_client
from claude_recall.storage.hot.redis_client import get_redis_client
from claude_recall.config import get_config

logger = logging.getLogger(__name__)


class QueryRouter:
    """
    Routes queries to the appropriate storage tier.

    Strategy:
    1. Try hot tier first (Redis) - ~1-5ms
    2. If insufficient results, query cold tier (PostgreSQL) - ~100-200ms
    3. Merge and deduplicate results
    """

    def __init__(self):
        postgres = get_postgres_client()
        redis = get_redis_client()

        # Search engines
        self.cold_observations = PostgresObservationSearch(postgres)
        self.cold_summaries = PostgresSummarySearch(postgres)
        self.hot_observations = RedisObservationSearch(redis)
        self.hot_summaries = RedisSummarySearch(redis)

        self.config = get_config()
        self._hot_available = True

    async def check_hot_tier(self) -> bool:
        """Check if hot tier is available."""
        try:
            redis = get_redis_client()
            self._hot_available = await redis.is_available()
        except Exception:
            self._hot_available = False
        return self._hot_available

    async def query(
        self,
        query_text: str,
        embedding: list[float],
        options: Optional[HybridSearchOptions] = None,
    ) -> RAGQueryResult:
        """
        Execute a tiered query for RAG context retrieval.

        Returns both observations and summaries relevant to the query.
        """
        opts = options or HybridSearchOptions()
        all_results: list[SearchResult] = []
        sources_used: list[str] = []
        hot_hits = 0
        cold_hits = 0

        # Hot tier search (if available)
        if self._hot_available:
            try:
                hot_results = await self.hot_observations.vector_search(embedding, opts)
                hot_hits = len(hot_results)
                all_results.extend(hot_results)
                if hot_hits > 0:
                    sources_used.append("hot")
                logger.debug(f"Hot tier returned {hot_hits} results")
            except Exception as e:
                logger.warning(f"Hot tier search failed: {e}")
                self._hot_available = False

        # Cold tier search (always, for completeness)
        min_hot_results = self.config.search.min_hot_results
        if not self._hot_available or hot_hits < min_hot_results:
            try:
                cold_opts = HybridSearchOptions(
                    limit=opts.limit,
                    project=opts.project,
                    since_epoch=opts.since_epoch,
                    types=opts.types,
                    min_score=opts.min_score,
                )
                cold_results = await self.cold_observations.hybrid_search(
                    query_text, embedding, cold_opts
                )
                cold_hits = len(cold_results)
                all_results.extend(cold_results)
                if cold_hits > 0:
                    sources_used.append("cold")
                logger.debug(f"Cold tier returned {cold_hits} results")
            except Exception as e:
                logger.error(f"Cold tier search failed: {e}")

        # Also search summaries for broader context
        summary_results = await self._search_summaries(query_text, embedding, opts)
        all_results.extend(summary_results)

        # Deduplicate and rank
        deduplicated = self._deduplicate_results(all_results)
        ranked = sorted(deduplicated, key=lambda x: x.score, reverse=True)

        # Apply final limit
        final_results = ranked[:opts.limit]

        # Update ranks
        for i, result in enumerate(final_results):
            result.rank = i + 1

        return RAGQueryResult(
            query=query_text,
            results=final_results,
            total_results=len(deduplicated),
            hot_tier_hits=hot_hits,
            cold_tier_hits=cold_hits,
            sources_used=sources_used,
        )

    async def _search_summaries(
        self,
        query_text: str,
        embedding: list[float],
        options: HybridSearchOptions,
    ) -> list[SearchResult]:
        """Search session summaries for context."""
        results: list[SearchResult] = []

        # Limit summary results to not overwhelm observations
        summary_limit = max(3, options.limit // 3)
        summary_opts = HybridSearchOptions(
            limit=summary_limit,
            project=options.project,
            since_epoch=options.since_epoch,
        )

        # Hot tier summaries
        if self._hot_available:
            try:
                hot_summaries = await self.hot_summaries.vector_search(
                    embedding, summary_opts
                )
                results.extend(hot_summaries)
            except Exception:
                pass

        # Cold tier summaries (hybrid search)
        try:
            cold_summaries = await self.cold_summaries.hybrid_search(
                query_text, embedding, summary_opts
            )
            results.extend(cold_summaries)
        except Exception as e:
            logger.warning(f"Summary search failed: {e}")

        return results

    def _deduplicate_results(
        self,
        results: list[SearchResult],
    ) -> list[SearchResult]:
        """Deduplicate results by ID and type, keeping highest score."""
        seen: dict[tuple[int, str], SearchResult] = {}

        for result in results:
            key = (result.id, result.type)
            if key not in seen or result.score > seen[key].score:
                seen[key] = result

        return list(seen.values())

    async def search_by_concepts(
        self,
        concepts: list[str],
        options: Optional[HybridSearchOptions] = None,
    ) -> list[SearchResult]:
        """Search by concept overlap (cold tier only)."""
        return await self.cold_observations.search_by_concepts(concepts, options)

    async def search_by_files(
        self,
        files: list[str],
        options: Optional[HybridSearchOptions] = None,
    ) -> list[SearchResult]:
        """Search by file involvement (cold tier only)."""
        return await self.cold_observations.search_by_files(files, options)


# Singleton instance
_router: Optional[QueryRouter] = None


def get_query_router() -> QueryRouter:
    """Get the global query router."""
    global _router
    if _router is None:
        _router = QueryRouter()
    return _router

File: src/services/storage/tiered/QueryRouter.ts
/**
 * Query Router
 * Routes search queries to appropriate tier(s) and merges results
 * Supports hot-first lookup with cold fallback for RAG retrieval
 */

import {
  getRedisObservationSearch,
  type RedisObservationSearch,
} from '../hot/RedisObservationStore.js';
import {
  getPostgresObservationSearch,
  getPostgresSummarySearch,
  type PostgresObservationSearch,
  type PostgresSummarySearch,
} from '../cold/PostgresHybridSearch.js';
import { getRedisClient } from '../hot/RedisClient.js';
import type { SearchResult, HybridSearchOptions } from '../interfaces/ISearchEngine.js';
import type { StoredObservation } from '../interfaces/IObservationStore.js';
import type { StoredSessionSummary } from '../interfaces/ISummaryStore.js';
import {
  getOllamaEmbeddingService,
  type OllamaEmbeddingService,
} from '../embedding/OllamaEmbedding.js';
import { getTieredStorageConfig } from '../config.js';

export interface RAGQueryOptions {
  query: string;
  project?: string;
  limit?: number;
  includeObservations?: boolean;
  includeSummaries?: boolean;
  minScore?: number;
  maxTokens?: number;
}

export interface RAGResult {
  observations: SearchResult<StoredObservation>[];
  summaries: SearchResult<StoredSessionSummary>[];
  queryTimeMs: number;
  hotTierHit: boolean;
}

export class QueryRouter {
  private hotObsSearch: RedisObservationSearch | null = null;
  private coldObsSearch: PostgresObservationSearch;
  private coldSummarySearch: PostgresSummarySearch;
  private embeddingService: OllamaEmbeddingService;
  private redisAvailable: boolean = false;

  constructor() {
    this.coldObsSearch = getPostgresObservationSearch();
    this.coldSummarySearch = getPostgresSummarySearch();
    this.embeddingService = getOllamaEmbeddingService();
  }

  /**
   * Initialize the query router
   */
  async initialize(): Promise<void> {
    try {
      const redisClient = getRedisClient();
      if (await redisClient.isAvailable()) {
        this.hotObsSearch = getRedisObservationSearch();
        this.redisAvailable = true;
        console.log('[QueryRouter] Redis hot tier search available');
      }
    } catch {
      this.redisAvailable = false;
    }
  }

  /**
   * Execute a RAG query with hot-first strategy
   *
   * Strategy:
   * 1. Generate embedding for query
   * 2. Try hot tier first (Redis) - fast ~1-5ms
   * 3. If insufficient results, query cold tier (PostgreSQL) - ~100-200ms
   * 4. Merge and deduplicate results
   */
  async queryForRAG(options: RAGQueryOptions): Promise<RAGResult> {
    const startTime = Date.now();
    const {
      query,
      project,
      limit = 10,
      includeObservations = true,
      includeSummaries = true,
      minScore = 0.3,
    } = options;

    const config = getTieredStorageConfig();
    const hotTierCutoff = Date.now() - (config.redis.hotTierTTL * 1000);

    // Generate embedding for query
    let embedding: number[] | null = null;
    try {
      if (await this.embeddingService.isAvailable()) {
        embedding = await this.embeddingService.generateEmbedding(query);
      }
    } catch (err) {
      console.warn('[QueryRouter] Failed to generate query embedding:', err);
    }

    const observations: SearchResult<StoredObservation>[] = [];
    const summaries: SearchResult<StoredSessionSummary>[] = [];
    let hotTierHit = false;

    // Query observations
    if (includeObservations) {
      // Try hot tier first
      if (this.redisAvailable && this.hotObsSearch && embedding) {
        try {
          const hotResults = await this.hotObsSearch.hybridSearch({
            query,
            embedding,
            limit,
            project,
            minScore,
          });

          if (hotResults.length >= limit * 0.7) {
            // Hot tier has enough results
            observations.push(...hotResults);
            hotTierHit = true;
          } else {
            // Need cold tier supplement
            const coldResults = await this.coldObsSearch.hybridSearch({
              query,
              embedding,
              limit: limit * 2,
              project,
              since_epoch: undefined, // Search all time in cold tier
            });

            // Merge and dedupe
            const hotIds = new Set(hotResults.map((r) => r.item.id));
            const deduped = coldResults.filter((r) => !hotIds.has(r.item.id));
            observations.push(...hotResults, ...deduped.slice(0, limit - hotResults.length));
          }
        } catch (err) {
          console.warn('[QueryRouter] Hot tier search failed:', err);
        }
      }

      // Fallback or supplement from cold tier
      if (observations.length < limit) {
        const needed = limit - observations.length;
        const existingIds = new Set(observations.map((r) => r.item.id));

        if (embedding) {
          const coldResults = await this.coldObsSearch.hybridSearch({
            query,
            embedding,
            limit: needed * 2,
            project,
            useReranker: config.search.useReranker,
          });

          const deduped = coldResults.filter((r) => !existingIds.has(r.item.id));
          observations.push(...deduped.slice(0, needed));
        } else {
          // Text-only fallback
          const textResults = await this.coldObsSearch.textSearch({
            query,
            limit: needed,
            project,
          });

          const deduped = textResults.filter((r) => !existingIds.has(r.item.id));
          observations.push(...deduped);
        }
      }
    }

    // Query summaries (cold tier only for now)
    if (includeSummaries && embedding) {
      const summaryResults = await this.coldSummarySearch.hybridSearch({
        query,
        embedding,
        limit: Math.ceil(limit / 2),
        project,
      });
      summaries.push(...summaryResults);
    }

    // Sort by score
    observations.sort((a, b) => b.score - a.score);
    summaries.sort((a, b) => b.score - a.score);

    return {
      observations: observations.slice(0, limit),
      summaries: summaries.slice(0, Math.ceil(limit / 2)),
      queryTimeMs: Date.now() - startTime,
      hotTierHit,
    };
  }

  /**
   * Quick search for recent relevant context (optimized for per-prompt RAG)
   */
  async quickSearch(
    query: string,
    project: string,
    limit: number = 5
  ): Promise<SearchResult<StoredObservation>[]> {
    // Generate embedding
    let embedding: number[] | null = null;
    try {
      if (await this.embeddingService.isAvailable()) {
        embedding = await this.embeddingService.generateEmbedding(query);
      }
    } catch {
      // Text-only fallback
    }

    // Hot tier only for quick search
    if (this.redisAvailable && this.hotObsSearch && embedding) {
      try {
        return await this.hotObsSearch.vectorSearch({
          embedding,
          limit,
          project,
        });
      } catch {
        // Fallback to cold
      }
    }

    // Cold tier fallback
    if (embedding) {
      return await this.coldObsSearch.vectorSearch({
        embedding,
        limit,
        project,
      });
    }

    // Text-only fallback
    return await this.coldObsSearch.textSearch({
      query,
      limit,
      project,
    });
  }

  /**
   * Format RAG results for context injection
   */
  formatForInjection(result: RAGResult, tokenBudget: number = 2000): string {
    const parts: string[] = [];
    const CHARS_PER_TOKEN = 4;
    let remainingChars = tokenBudget * CHARS_PER_TOKEN;

    // Add observations (higher priority)
    if (result.observations.length > 0) {
      parts.push('## Relevant Past Work\n');
      remainingChars -= 25;

      for (const { item, score } of result.observations) {
        if (remainingChars <= 0) break;

        const obsText = this.formatObservation(item, score);
        if (obsText.length <= remainingChars) {
          parts.push(obsText);
          remainingChars -= obsText.length;
        }
      }
    }

    // Add summaries
    if (result.summaries.length > 0 && remainingChars > 200) {
      parts.push('\n## Recent Session Summaries\n');
      remainingChars -= 30;

      for (const { item, score } of result.summaries) {
        if (remainingChars <= 0) break;

        const summaryText = this.formatSummary(item, score);
        if (summaryText.length <= remainingChars) {
          parts.push(summaryText);
          remainingChars -= summaryText.length;
        }
      }
    }

    return parts.join('\n');
  }

  private formatObservation(obs: StoredObservation, score: number): string {
    const parts: string[] = [];

    // Type and title
    parts.push(`### [${obs.type}] ${obs.title || 'Untitled'}`);

    // Subtitle if present
    if (obs.subtitle) {
      parts.push(`_${obs.subtitle}_`);
    }

    // Narrative or facts
    if (obs.narrative) {
      parts.push(obs.narrative);
    } else if (obs.facts && obs.facts.length > 0) {
      parts.push(obs.facts.map((f) => `- ${f}`).join('\n'));
    }

    // Files if present
    if (obs.files_modified && obs.files_modified.length > 0) {
      parts.push(`Files: ${obs.files_modified.slice(0, 3).join(', ')}`);
    }

    parts.push('');
    return parts.join('\n');
  }

  private formatSummary(summary: StoredSessionSummary, score: number): string {
    const parts: string[] = [];

    if (summary.request) {
      parts.push(`**Request:** ${summary.request}`);
    }
    if (summary.completed) {
      parts.push(`**Completed:** ${summary.completed}`);
    }
    if (summary.learned) {
      parts.push(`**Learned:** ${summary.learned}`);
    }

    parts.push('');
    return parts.join('\n');
  }
}

// Singleton instance
let queryRouterInstance: QueryRouter | null = null;

export function getQueryRouter(): QueryRouter {
  if (!queryRouterInstance) {
    queryRouterInstance = new QueryRouter();
  }
  return queryRouterInstance;
}

export async function initializeQueryRouter(): Promise<QueryRouter> {
  const router = getQueryRouter();
  await router.initialize();
  return router;
}
